pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,283 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	gcpclient "github.com/gardener/gardener-extension-provider-gcp/pkg/internal/client"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	ctrlerror "github.com/gardener/gardener/extensions/pkg/controller/error"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	"google.golang.org/api/compute/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "reconcile")
+
+	gcpclient, err := a.getGCPClient(ctx, bastion)
+	if err != nil {
+		return fmt.Errorf("%w, failed to create GCP client", err)
+	}
+
+	opt, err := DetermineOptions(ctx, bastion, cluster)
+	if err != nil {
+		return fmt.Errorf("%w, failed to setup GCP options", err)
+	}
+
+	err = ensureFirewallRule(ctx, gcpclient, opt)
+	if err != nil {
+		return fmt.Errorf("%w, failed to ensure firewall rule", err)
+	}
+
+	endpoints, err := ensureBastionInstance(ctx, logger, bastion, gcpclient, opt)
+	if err != nil {
+		return fmt.Errorf("%w, failed to ensure bastion instance", err)
+	}
+
+	if !endpoints.Ready() {
+		return &ctrlerror.RequeueAfterError{
+			// requeue rather soon, so that the user (most likely gardenctl eventually)
+			// doesn't have to wait too long for the public endpoint to become available
+			RequeueAfter: 5 * time.Second,
+			Cause:        fmt.Errorf("bastion instance has no public/private endpoints yet"),
+		}
+	}
+
+	// once a public endpoint is available, publish the endpoint on the
+	// Bastion resource to notify upstream about the ready instance
+	return controller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.Client(), bastion, func() error {
+		bastion.Status.Ingress = *endpoints.public
+		return nil
+	})
+
+}
+
+func ensureFirewallRule(ctx context.Context, gcpclient gcpclient.Interface, opt *Options) error {
+	firewall, err := getFirewallRule(ctx, gcpclient, opt)
+	if err != nil {
+		return fmt.Errorf("%w, could not get firewall rule", err)
+	}
+
+	// create firewall if it doesn't exist yet
+	if firewall == nil {
+		return createFirewallRule(ctx, gcpclient, opt)
+	}
+
+	a := firewall.SourceRanges
+	b := opt.PublicIP
```

Comment:
use better names instead of `a` and `b`

```go
currentCIDRs := firewall.SourceRanges
wantedCIDRs := opt.PublicIP
```

Reply:
committed in c1472d2ede0f5d6334526fa6fb08c4e719c43b4b